rust/agama-lib/src/base_http_client.rs

```diff
@@ -109,48 +112,108 @@ impl BaseHTTPClient {
+    pub async fn post_response(
+        &self,
+        path: &str,
+        object: &impl Serialize,
+    ) -> Result<Response, ServiceError> {
+        self.request_response(reqwest::Method::POST, path, object)
+            .await
+    }
+
+    /// put object to given path, deserializes the response
+    ///
+    /// Arguments:
+    ///
+    /// * `path`: path relative to HTTP API like `/users/first`
+    /// * `object`: Object that can be serialiazed to JSON as body of request.
+    pub async fn put<T>(&self, path: &str, object: &impl Serialize) -> Result<T, ServiceError>
+    where
+        T: DeserializeOwned,
+    {
+        let response = self
+            .request_response(reqwest::Method::PUT, path, object)
+            .await?;
         if response.status().is_success() {
-            Ok(())
+            response.json::<T>().await.map_err(|e| e.into())
         } else {
             Err(self.build_backend_error(response).await)
         }
     }
 
-    /// post object to given path and returns server response. Reports error only if failed to send
+    /// put object to given path and report error if response is not success
+    ///
+    /// Arguments:
+    ///
+    /// * `path`: path relative to HTTP API like `/users/first`
+    /// * `object`: Object that can be serialiazed to JSON as body of request.
+    pub async fn put_void(&self, path: &str, object: &impl Serialize) -> Result<(), ServiceError> {
+        let response = self
+            .request_response(reqwest::Method::PUT, path, object)
+            .await?;
+        self.unit_or_error(response).await
+    }
+
+    /// POST/PUT/PATCH an object to a given path and returns server response.
+    /// Reports Err only if failed to send
     /// request, but if server returns e.g. 500, it will be in Ok result.
     ///
     /// In general unless specific response handling is needed, simple post should be used.
     ///
     /// Arguments:
     ///
+    /// * `method`: for example `reqwest::Method::PUT`
     /// * `path`: path relative to HTTP API like `/questions`
     /// * `object`: Object that can be serialiazed to JSON as body of request.
-    pub async fn post_response(
+    pub async fn request_response(
         &self,
+        method: reqwest::Method,
         path: &str,
         object: &impl Serialize,
     ) -> Result<Response, ServiceError> {
         self.client
-            .post(self.url(path))
+            .request(method, self.url(path))
             .json(object)
             .send()
             .await
             .map_err(|e| e.into())
     }
 
+    /// patch object at given path and report error if response is not success
+    ///
+    /// Arguments:
+    ///
+    /// * `path`: path relative to HTTP API like `/users/first`
+    /// * `object`: Object that can be serialiazed to JSON as body of request.
+    pub async fn patch(&self, path: &str, object: &impl Serialize) -> Result<(), ServiceError> {
```

Comment:
I would expect a `patch_response` method too. Like `PATCH`, nothing prevents a response to a `PATCH` request from having a body.

Reply:
Right. Have `patch` and `patch_void` now.